### PR TITLE
implement initial version of v4 signing for GCP

### PR DESCRIPTION
**Pull request by wolfv**
This is by far not done, but works in my manual tests :)

- Google advertises the RSA method as method of choice and it seems that it is more natural, but it adds another dependency (`rsa`). Should we put it behind a feature?
- Also needed serde to deserialize the json files one gets from Google for the RSA key.
- Google also supports `amz-...` functions but everything must be prefixed with that then. Should we factor out the HMAC functions and make them work for both aws and google?

Haven't implemented the proper config loader yet! Will try to get this done over the coming days.

### CODE REVIEW

**Comment by Xuanwo** [comment]
Since google also supports AWS V4 signing, how about using `AwsV4Signer` in `gcs` service directly? Seem we only need to allow users access the `client_email` and `private_key` (and use them as ak/sk)?

**Comment by wolfv** [comment]
Yeah, that would probably work, but we wouldn't be able to use the "native" credentials (only those obtained for interoperability. The native ones use RSA for the signature.

Do you know how you want to have the API to support both Google authentication methods? Happy to follow your advice (I only exported them so that I could test with an example easily).

**Comment by Xuanwo** [comment]
> Do you know how you want to have the API to support both Google authentication methods?

No feature requests so far. I prefer to let gcs query sign work first~

**Comment by Xuanwo** [comment]
Please update branch to address the unexpected panic, sorry for the unexpected CI broken.

File: src/google/signer.rs [hunk 1]
```diff
@@ -213,6 +217,40 @@ impl Signer {
...
+        req: &impl SignableRequest,
+        method: SigningMethod,
+        cred: &Credential,
+    ) -> Result<CanonicalRequest> {
+        let mut creq = CanonicalRequest::new(
+            req,
+            method,
+            None,
+            SigningAlgorithm::Goog4Rsa,
+            "europe-west3".to_string(),
```
Comment (Xuanwo) [inline src/google/signer.rs:232]: Hardcoded？

File: src/google/signer.rs [hunk 2]
```diff
@@ -213,6 +217,40 @@ impl Signer {
...
+            "storage".to_string(),
+        )?;
+        creq.build_headers(cred)?;
+        creq.build_query(cred)?;
+
+        debug!("calculated canonical request: {creq}");
+        Ok(creq)
+    }
+
+    pub fn sign_v4(&self, req: &mut impl SignableRequest) -> Result<()> {
```
Comment (Xuanwo) [inline src/google/signer.rs:242]: `sign_v4` is the implementation detail that I don't want to expose to users. Please keep the same API with other signers like `sign_query`.

By the way, we don't need an `expire`?

File: src/google/signer.rs [hunk 3]
```diff
@@ -213,6 +217,40 @@ impl Signer {
...
+        let credential = self
+            .credential_loader
+            .load_credential()
+            .expect("Did not find credentials for google.");
+        let v4cred: Credential = credential.into();
```
Comment (Xuanwo) [inline src/google/signer.rs:247]: I'm not a fan of `From/Into` API, how about create `AwsCredential` here directly? It looks more clear to me.

File: src/google/signer.rs [hunk 4]
```diff
@@ -213,6 +217,40 @@ impl Signer {
...
+        let mut creq = self.canonicalize(req, SigningMethod::Header, &v4cred)?;
+        println!("Canonicalized!");
```
Comment (Xuanwo) [inline src/google/signer.rs:249]: Please use `debug`

File: src/google/signer.rs [hunk 5]
```diff
@@ -213,6 +217,40 @@ impl Signer {
...
+        Ok(creq)
+    }
+
+    pub fn sign_v4(&self, req: &mut impl SignableRequest) -> Result<()> {
+        let credential = self
+            .credential_loader
+            .load_credential()
+            .expect("Did not find credentials for google.");
+        let v4cred: Credential = credential.into();
+        let mut creq = self.canonicalize(req, SigningMethod::Header, &v4cred)?;
```
Comment (Xuanwo) [inline src/google/signer.rs:248]: I guess we need to use `query`?

File: src/google/signer.rs (hunk 1 above)
Comment (wolfv) [reply src/google/signer.rs:232]: I am still testing things, sorry for the noise :)

File: src/google/signer.rs (hunk 2 above)
Comment (wolfv) [reply src/google/signer.rs:242]: And how do you want to choose between the previous way and v4? We can decide HMAC vs RSA based on credential.

Comment (Xuanwo) [reply src/google/signer.rs:242]: > And how do you want to choose between the previous way and v4? We can decide HMAC vs RSA based on credential.

No need to choose at this point. We use bearer token for `sign` and aws v4 for `sign_query`. Other features like gcs's ras based key can be added in the future if we have a request.

Comment (wolfv) [reply src/google/signer.rs:242]: Hmm, I do think that we "request" this feature and would like to use it. But we can start with only AWS HMAC if you prefer.

**Comment by wolfv** [comment]
How should we get HMAC keys for google? Using `AWS_SECRET_ACCESS_KEY -> GOOGLE_SECRET_ACCESS_KEY` ? Or `GCP_SECRET...`? 

Not sure if there is a well-defined default env name for these

File: src/google/signer.rs (hunk 5 above)
Comment (wolfv) [reply src/google/signer.rs]: I've implemented `sign_query` now and use the `Query` method, yes.

**Comment by Xuanwo** [comment]
> How should we get HMAC keys for google? Using `AWS_SECRET_ACCESS_KEY -> GOOGLE_SECRET_ACCESS_KEY` ? Or `GCP_SECRET...`?

We can use `AwsCredential::new(credential.client_email(), credential.private_key());` as the key to sign query?

File: src/google/signer.rs (hunk 2 above)
Comment (Xuanwo) [reply src/google/signer.rs]: Oh, sorry. I must miss your feature request. The only feature I remember for is we want OpenDAL to support gcs query sign. 

And I assume that you don't care about the detailed implementation that reqsign use underhood. **Please correct me if I understand incorrectly**. The following design is based on my previous assumption.

So the best option to me is:

- Use AwsSigner to do query sign for gcs. (Please let know if this way doesn't work)
- No other changes (like adding `SigningMethod` or `SigningFlavor`, I don't want to reuse code in this way)

**Comment by wolfv** [comment]
That only works for the RSA signature because that key in the Google case is an RSA key. HMAC is only provided for compatibility and not as the JSON that is usually provided for authentication as far as I know.

**Comment by wolfv** [comment]
Hey @Xuanwo thanks for the thumbs up but idk how to proceed :) you didn't want RSA so for HMAC we need to load the Google HMAC credentials somehow. I'll use the env var as proposed?

**Comment by Xuanwo** [comment]
> Hey @Xuanwo thanks for the thumbs up but idk how to proceed :) you didn't want RSA so for HMAC we need to load the Google HMAC credentials somehow. I'll use the env var as proposed?

Sorry for my misunderstanding.

Now I get it:

Google supports the following three ways to do query sign: aws hmac (s3 compatible), google hmac (similar to s3, but owned by google) and google rsa. Google ras key can be extracted from existing credentials. Also, google provides hmac key which be used both in aws hmac and google hmac. (Please correct me, thanks)

So my current idea is implement pure google rsa support, so our user can use existing credentials directly without add new settings. 

For code layout, please copy useful code from aws service, I'm planning a future refactor to use them in a more general way.

Looks good to you? Sorry again.

**Comment by wolfv** [comment]
Sounds good to me! I thought it was not desirable because it brings a new dependency (`rsa`).

Even for RSA we need the same CanonicalRequest. I already factored it out in this PR. Does that look OK to you or how do you want it factored out so that we can use it both from AWS and Google?

**Comment by Xuanwo** [comment]
> Does that look OK to you or how do you want it factored out so that we can use it both from AWS and Google?

It's by design to keep them separated for now. Please feel free to copy code related `CanonicalRequest`.

Other services like azblob, obs also have similar logic, so it's not a good idea to do it now in this PR.

**Comment by wolfv** [comment]
hmm too bad, because for google and AWS it is _exactly_ the same code except for GOOG4 vs AWS4 prefixes ...

**Comment by Xuanwo** [comment]
> hmm too bad, because for google and AWS it is _exactly_ the same code except for GOOG4 vs AWS4 prefixes ...

Don't worry, they will be refactor later.

**Comment by wolfv** [comment]
continuing in #247

**Pull request closed without merging**